{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: 'gplots'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following object is masked from 'package:limma':\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:gplots':\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grDevices':\n",
      "\n",
      "    windows\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in library(RcolorBrewer): there is no package called 'RcolorBrewer'\n",
     "output_type": "error",
     "traceback": [
      "Error in library(RcolorBrewer): there is no package called 'RcolorBrewer'\nTraceback:\n",
      "1. library(RcolorBrewer)"
     ]
    }
   ],
   "source": [
    "# load in libraries\n",
    "library(edgeR)\n",
    "library(limma)\n",
    "library(Glimma)\n",
    "library(gplots)\n",
    "library(org.Mm.eg.db)\n",
    "library(RColorBrewer)\n",
    "\n",
    "# count matrix\n",
    "seqdata <- read.delim(\"data/GSE60450_Lactation-GenewiseCounts.txt\")\n",
    "head(seqdata)\n",
    "dim(seqdata)\n",
    "\n",
    "# sample information\n",
    "sampleinfo <- read.delim(\"data/SampleInfo.txt\")\n",
    "dim(sampleinfo)\n",
    "head(sampleinfo)\n",
    "\n",
    "head(seqdata)\n",
    "countdata <- seqdata[, -c(1, 2)]\n",
    "head(countdata)\n",
    "# assign gene ids\n",
    "rownames(countdata) <- seqdata[, 1]\n",
    "\n",
    "head(countdata)\n",
    "\n",
    "rownames(countdata)[1:10]\n",
    "colnames(countdata)\n",
    "\n",
    "colnames(countdata) <- substr(colnames(countdata), 1, 7)\n",
    "head(countdata)\n",
    "\n",
    "sampleinfo\n",
    "colnames(countdata)\n",
    "\n",
    "sampleinfo$SampleName == colnames(countdata)\n",
    "table(sampleinfo$SampleName == colnames(countdata))\n",
    "\n",
    "# filtering lowly expressed genes\n",
    "myCPM <- cpm(countdata)\n",
    "head(myCPM)\n",
    "\n",
    "thresh <- myCPM > 0.5\n",
    "head(thresh)\n",
    "\n",
    "TRUE + TRUE\n",
    "table(rowSums(thresh))\n",
    "\n",
    "# genes to keep\n",
    "keep <- rowSums(thresh) >= 2\n",
    "\n",
    "count.keep <- countdata[keep, ]\n",
    "dim(countdata)\n",
    "dim(count.keep)\n",
    "\n",
    "table(keep)\n",
    "\n",
    "# plot cpm vs counts\n",
    "plot(myCPM[, 1], countdata[, 1])\n",
    "abline(v = 0.5)\n",
    "\n",
    "plot(myCPM[, 1], countdata[, 1], ylim = c(0, 50), xlim = c(0, 3))\n",
    "abline(v = 0.5)\n",
    "\n",
    "# PLOTTING CODE (Part 2):\n",
    "    # load in libraries\n",
    "library(edgeR)\n",
    "library(limma)\n",
    "library(Glimma)\n",
    "library(gplots)\n",
    "library(org.Mm.eg.db)\n",
    "library(RColorBrewer)\n",
    "\n",
    "# count matrix\n",
    "seqdata <- read.delim(\"data/GSE60450_Lactation-GenewiseCounts.txt\")\n",
    "head(seqdata)\n",
    "dim(seqdata)\n",
    "\n",
    "# sample information\n",
    "sampleinfo <- read.delim(\"data/SampleInfo.txt\")\n",
    "dim(sampleinfo)\n",
    "head(sampleinfo)\n",
    "\n",
    "head(seqdata)\n",
    "countdata <- seqdata[, -c(1, 2)]\n",
    "head(countdata)\n",
    "# assign gene ids\n",
    "rownames(countdata) <- seqdata[, 1]\n",
    "\n",
    "head(countdata)\n",
    "\n",
    "rownames(countdata)[1:10]\n",
    "colnames(countdata)\n",
    "\n",
    "colnames(countdata) <- substr(colnames(countdata), 1, 7)\n",
    "head(countdata)\n",
    "\n",
    "sampleinfo\n",
    "colnames(countdata)\n",
    "\n",
    "sampleinfo$SampleName == colnames(countdata)\n",
    "table(sampleinfo$SampleName == colnames(countdata))\n",
    "\n",
    "# filtering lowly expressed genes\n",
    "myCPM <- cpm(countdata)\n",
    "head(myCPM)\n",
    "\n",
    "thresh <- myCPM > 0.5\n",
    "head(thresh)\n",
    "\n",
    "TRUE + TRUE\n",
    "table(rowSums(thresh))\n",
    "\n",
    "# genes to keep\n",
    "keep <- rowSums(thresh) >= 2\n",
    "\n",
    "count.keep <- countdata[keep, ]\n",
    "dim(countdata)\n",
    "dim(count.keep)\n",
    "\n",
    "table(keep)\n",
    "\n",
    "# plot cpm vs counts\n",
    "plot(myCPM[, 1], countdata[, 1])\n",
    "abline(v = 0.5)\n",
    "\n",
    "plot(myCPM[, 1], countdata[, 1], ylim = c(0, 50), xlim = c(0, 3))\n",
    "abline(v = 0.5)\n",
    "\n",
    "# Challenge\n",
    "plot(myCPM[, 2], countdata[, 2], ylim = c(0, 50), xlim = c(0, 3))\n",
    "abline(v = 0.5, col = \"blue\")\n",
    "abline(h = 10, col = \"blue\")\n",
    "\n",
    "# creating a DGEList object\n",
    "sampleinfo\n",
    "head(count.keep)\n",
    "y <- DGEList(count.keep)\n",
    "\n",
    "class(y)\n",
    "names(y)\n",
    "y\n",
    "head(y$count)\n",
    "head(y$samples)\n",
    "\n",
    "y$samples$lib.size\n",
    "# barplot of library sizes\n",
    "barplot(y$samples$lib.size, names = colnames(y), las = 2)\n",
    "title(\"Barplot of library sizes\")\n",
    "\n",
    "# boxplot of log counts\n",
    "logcounts <- cpm(y, log=TRUE)\n",
    "boxplot(logcounts, xlab = \"\", ylab = \"Log2 counts per million\", las = 2)\n",
    "abline(h = median(logcounts), col = \"blue\")\n",
    "title(\"Boxplots of logCPMs(unnormalised)\")\n",
    "\n",
    "# MDS plot\n",
    "plotMDS(y)\n",
    "\n",
    "sampleinfo$CellType\n",
    "class(sampleinfo$CellType)\n",
    "as.numeric(sampleinfo$CellType)\n",
    "col.cell <- c(\"purple\", \"orange\")[sampleinfo$CellType]\n",
    "\n",
    "data.frame(sampleinfo$CellType, col.cell)\n",
    "\n",
    "# colour by cell type\n",
    "plotMDS(y, col = col.cell)\n",
    "legend(\"topleft\", fill = c(\"purple\", \"orange\"), legend = levels(sampleinfo$CellType))\n",
    "\n",
    "# colour by status\n",
    "col.status <- c(\"blue\", \"red\", \"dark green\")[sampleinfo$Status]\n",
    "plotMDS(y, col = col.status)\n",
    "\n",
    "# plot different colourings side by side\n",
    "par(mfrow = c(1, 2))\n",
    "plotMDS(y, col = col.cell)\n",
    "plotMDS(y, col = col.status)\n",
    "\n",
    "# load in corrected sample annotation\n",
    "# previously shown that samples had labels switched\n",
    "sampleinfo <- read.delim(\"data/SampleInfo_Corrected.txt\")\n",
    "\n",
    "col.cell <- c(\"purple\", \"orange\")[sampleinfo$CellType]\n",
    "col.status <- c(\"blue\", \"red\", \"dark green\")[sampleinfo$Status]\n",
    "\n",
    "# plot again to see corrected annotation\n",
    "par(mfrow = c(1, 2))\n",
    "plotMDS(y, col = col.cell)\n",
    "legend(\"topleft\", fill = c(\"purple\", \"orange\"), legend = levels(sampleinfo$CellType))\n",
    "plotMDS(y, col = col.status)\n",
    "legend(\"topleft\", fill = c(\"blue\",\"red\",\"dark green\"), legend = levels(sampleinfo$Status), cex=0.8)\n",
    "title(\"Status\")\n",
    "\n",
    "# interactive version of plot\n",
    "glMDSPlot(y, group = sampleinfo[, c(\"CellType\", \"Status\")])\n",
    "\n",
    "# reset the settings so plots take up whole window\n",
    "par(mfrow = c(1, 1))\n",
    "\n",
    "# Challenge\n",
    "col.status <- c(\"salmon\", \"cornflowerblue\", \"mediumvioletred\")[sampleinfo$Status]\n",
    "pch.cell <- c(1, 4)[sampleinfo$CellType]\n",
    "\n",
    "plotMDS(y, col = col.status, pch = pch.cell)\n",
    "legend(\"topleft\", fill = c(\"salmon\", \"cornflowerblue\", \"mediumvioletred\"), legend = levels(sampleinfo$Status), cex=0.8)\n",
    "legend(\"bottom\", pch = c(1, 4), legend = levels(sampleinfo$CellType), cex=0.8)\n",
    "\n",
    "# heatmap\n",
    "var_genes <- apply(logcounts, MARGIN = 1, FUN = var)\n",
    "\n",
    "select_var <- names( sort(var_genes, decreasing = TRUE) )[1:500]\n",
    "\n",
    "names(head( sort(var_genes, decreasing = TRUE) ))\n",
    "\n",
    "head(select_var)\n",
    "\n",
    "highly_variable_lcpm <- logcounts[select_var, ]\n",
    "dim(highly_variable_lcpm)\n",
    "\n",
    "mypalette <- brewer.pal(11, \"RdYlBu\")\n",
    "morecols <- colorRampPalette(mypalette)\n",
    "\n",
    "col.cell <- c(\"purple\", \"orange\")[sampleinfo$CellType]\n",
    "\n",
    "heatmap.2(\n",
    "    highly_variable_lcpm,\n",
    "    col = rev(morecols(50)),\n",
    "    trace = \"none\",\n",
    "    main = \"Top 500 most variable genes across samples\",\n",
    "    ColSideColors = col.cell,\n",
    "    scale = \"row\"\n",
    ")\n",
    "\n",
    "# Challenge Part 1\n",
    "mypalette <- brewer.pal(11, \"PiYG\")\n",
    "morecols <- colorRampPalette(mypalette)\n",
    "\n",
    "column_labels <- paste(sampleinfo$CellType, sampleinfo$Status,\n",
    "    sep = \".\")\n",
    "\n",
    "heatmap.2(\n",
    "    highly_variable_lcpm,\n",
    "    col = rev(morecols(50)),\n",
    "    trace = \"none\",\n",
    "    main = \"Top 500 most variable genes across samples\",\n",
    "    ColSideColors = col.cell,\n",
    "    scale = \"row\",\n",
    "    labCol = column_labels,\n",
    "    labRow = \"\",\n",
    "    margins = c(9, 5)\n",
    ")\n",
    "\n",
    "# Challenge Part 3\n",
    "select_var <- names(sort(var_genes, decreasing = FALSE))[1:500]\n",
    "\n",
    "lowly_variable_lcpm <- logcounts[select_var, ]\n",
    "\n",
    "mypalette <- brewer.pal(11, \"PiYG\")\n",
    "morecols <- colorRampPalette(mypalette)\n",
    "\n",
    "col.cell <- c(\"purple\", \"orange\")[sampleinfo$CellType]\n",
    "\n",
    "lab_cols <- paste(sampleinfo$CellType, sampleinfo$Status,\n",
    "    sep = \".\")\n",
    "\n",
    "heatmap.2(\n",
    "    lowly_variable_lcpm,\n",
    "    col = rev(morecols(50)),\n",
    "    trace = \"none\",\n",
    "    main = \"Top 500 most variable genes across samples\",\n",
    "    ColSideColors = col.cell,\n",
    "    scale = \"row\",\n",
    "    labCol = lab_cols\n",
    ")\n",
    "\n",
    "# Normalisation for composition bias\n",
    "y <- calcNormFactors(y)\n",
    "y$samples\n",
    "\n",
    "par(mfrow = c(1, 2))\n",
    "plotMD(logcounts, column = 7)\n",
    "abline(h = 0, col = \"grey\")\n",
    "plotMD(logcounts, column = 11)\n",
    "abline(h = 0, col = \"grey\")\n",
    "\n",
    "plotMD(y, column = 7)\n",
    "abline(h = 0, col = \"grey\")\n",
    "plotMD(y, column = 11)\n",
    "abline(h = 0, col = \"grey\")\n",
    "\n",
    "\n",
    "# Code from Day 2\n",
    "print(y)\n",
    "group\n",
    "group <- paste(sampleinfo$CellType, sampleinfo$Status,sep=\".\")\n",
    "group <- factor(group)\n",
    "head(logcounts)\n",
    "\n",
    "# Let's set up our design matrix\n",
    "design <- model.matrix(~ 0 + group)\n",
    "design\n",
    "\n",
    "# We can parameterize models in different ways\n",
    "# This design includes an intercept, but that can be \n",
    "# complicated to interpret in experiments such as this!\n",
    "design2 <- model.matrix(~group)\n",
    "\n",
    "# Give our design matrix nicer column names\n",
    "colnames(design) <- levels(group)\n",
    "\n",
    "# Voom transform the data\n",
    "par(mfrow=c(1,1))\n",
    "v <- voom(y, design, plot=TRUE)\n",
    "?voom\n",
    "\n",
    "# Challenge\n",
    "# What is a targets slot of v? What does it correspond to in y?\n",
    "# What are the dimensions of the weights slot in v?\n",
    "\n",
    "# The weights are used in voom to deal with the mean-variance relationship\n",
    "# v$E are our transformed expression values that we'll use in our modelling\n",
    "\n",
    "dim(v$weights)\n",
    "dim(v)\n",
    "\n",
    "# Boxplot of data before and after compositional TMM norm\n",
    "par(mfrow=c(1,2))\n",
    "boxplot(logcounts, xlab=\"\", ylab=\"Log2 CPM\", las=2,\n",
    "        main=\"Library size + Log2 data\")\n",
    "abline(h=median(logcounts), col=\"blue\")\n",
    "boxplot(v$E, xlab=\"\", ylab=\"Log2 CPM\", las=2,\n",
    "        main=\"Library size, TMM, Log2 data\")\n",
    "abline(h=median(v$E), col=\"blue\")\n",
    "\n",
    "# Let's fit a linear model\n",
    "fit <- lmFit(v)\n",
    "class(fit)\n",
    "names(fit)\n",
    "?lmFit\n",
    "head(fit$coefficients)\n",
    "\n",
    "# Let's calculate relative expression changes log2FC\n",
    "# Define a contrast matrix\n",
    "cont.matrix <- makeContrasts(\n",
    "  B.PregVsLac=basal.pregnant-basal.lactate, \n",
    "  levels=design\n",
    ")\n",
    "\n",
    "cont.matrix\n",
    "\n",
    "# Let's estimate our contrast\n",
    "fit.cont <- contrasts.fit(fit, cont.matrix)\n",
    "\n",
    "# Shrinkage of gene-specific variances\n",
    "fit.cont <- eBayes(fit.cont)\n",
    "\n",
    "dim(fit.cont)\n",
    "\n",
    "# How much differential expression is there for \n",
    "# basal.preg vs basal.lact?\n",
    "summa.fit <- decideTests(fit.cont)\n",
    "summary(summa.fit)\n",
    "\n",
    "\n",
    "cont.matrix2 <- makeContrasts(\n",
    "  B.PregVsLac=basal.pregnant-basal.lactate, \n",
    "  L.PregVsLac=luminal.pregnant-luminal.lactate,\n",
    "  levels=design\n",
    ")\n",
    "\n",
    "cont.matrix2\n",
    "\n",
    "# Test for DE of both contrasts\n",
    "fit.cont2 <- contrasts.fit(fit, cont.matrix2)\n",
    "fit.cont2 <- eBayes(fit.cont2)\n",
    "dim(fit.cont2)\n",
    "\n",
    "summa.fit2 <- decideTests(fit.cont2)\n",
    "summary(summa.fit2)\n",
    "\n",
    "# Make a vennDiagram of the overlap\n",
    "par(mfrow=c(1,1))\n",
    "vennDiagram(summa.fit2,\n",
    "            include=c(\"up\", \"down\"))\n",
    "\n",
    "?decideTests\n",
    "\n",
    "fit.cont <- fit.cont2\n",
    "summa.fit <- summa.fit2\n",
    "\n",
    "# Rank genes by evidence for Differential expression\n",
    "topTable(fit.cont, sort.by='p')\n",
    "topTable(fit.cont, coef=1, sort.by='p')\n",
    "topTable(fit.cont, coef=\"B.PregVsLac\", sort.by='p')\n",
    "topTable(fit.cont, coef=\"L.PregVsLac\", sort.by='p')\n",
    "\n",
    "# Annotate results with gene names\n",
    "library(org.Mm.eg.db)\n",
    "columns(org.Mm.eg.db)\n",
    "\n",
    "ann <- select(org.Mm.eg.db, keys=rownames(fit.cont),\n",
    "              columns=c(\"ENTREZID\", \"SYMBOL\", \"GENENAME\"))\n",
    "dim(ann)\n",
    "head(ann)\n",
    "\n",
    "# Check that the order is correct! Most important step!\n",
    "table(ann$ENTREZID==rownames(fit.cont))\n",
    "\n",
    "# Assigned gene information to our fitted model object\n",
    "fit.cont$genes <- ann\n",
    "# Results now have gene annotation\n",
    "topTable(fit.cont, coef=\"B.PregVsLac\", sort.by='p')\n",
    "\n",
    "limma.res <- topTable(fit.cont, coef=\"B.PregVsLac\", \n",
    "                      number=Inf, sort.by='p')\n",
    "dim(limma.res)\n",
    "\n",
    "write.csv(limma.res, file=\"B.PregVsLacResults.csv\", row.names=FALSE)\n",
    "\n",
    "\n",
    "# Plotting of diffential expression results\n",
    "par(mfrow=c(1,1))\n",
    "plotMD(fit.cont, coef=1, status=summa.fit[,\"B.PregVsLac\"], \n",
    "       values=c(-1,1))\n",
    "volcanoplot(fit.cont, coef=1, highlight=100, \n",
    "            names=fit.cont$genes$SYMBOL)\n",
    "\n",
    "# Challenge: Repeat our plotting of results for L.PregVsLac contrast\n",
    "# Make a MD plot and a volcanoplot. \n",
    "# For the the volcanoplot highlight the top 200 genes\n",
    "par(mfrow=c(1,2))\n",
    "plotMD(fit.cont, coef=2, status=summa.fit[,2], \n",
    "       values=c(-1,1))\n",
    "volcanoplot(fit.cont, coef=2, highlight=200, \n",
    "            names=fit.cont$genes$SYMBOL)\n",
    "title(\"L.PregVsLac volcano\")\n",
    "\n",
    "par(mfrow=c(1,3))\n",
    "stripchart(v$E[\"24117\",]~group)\n",
    "stripchart(v$E[\"24117\",]~group, vertical=TRUE, las=2, \n",
    "           cex.axis=0.8, pch=16, col=1:6, method=\"jitter\")\n",
    "nice.col = brewer.pal(6, name='Dark2')\n",
    "stripchart(v$E[\"24117\",]~group, vertical=TRUE, las=2, \n",
    "           cex.axis=0.8, pch=16, col=nice.col, method=\"jitter\",\n",
    "           ylab=\"Normalised log2 Expression\", main=\"Wif1\")\n",
    "\n",
    "\n",
    "# Challenge. Repeat stip chart for top DE gene in L.PregVsLac contrast\n",
    "topTable(fit.cont, coef=2)\n",
    "stripchart(v$E[\"12992\",]~group, vertical=TRUE, las=2, \n",
    "           cex.axis=0.8, pch=16, col=nice.col, method=\"jitter\",\n",
    "           ylab=\"Normalised log2 Expression\", main=\"Csn1s2b\")\n",
    "\n",
    "\n",
    "# Let's make an interactive volcano plot\n",
    "group2 <- group\n",
    "levels(group2) <- c(\"basal.lactate\",  \"basal.preg\",  \n",
    "                    \"basal.virgin\", \"lum.lactate\", \n",
    "                    \"lum.preg\", \"lum.virgin\")\n",
    "\n",
    "glXYPlot(x=fit.cont$coefficients[,1], y=fit.cont$lods[,1],\n",
    "         xlab=\"logFC\", ylab=\"B\", main=\"B.PregVsLac\",\n",
    "         counts=v$E, groups=group2, status=summa.fit[,1],\n",
    "         anno=fit.cont$genes, side.main=\"ENTREZID\", folder=\"volcano\")\n",
    "\n",
    "glMDPlot(fit.cont, coef=1, main=\"B.PregVsLac\",\n",
    "         counts=v$E, groups=group2, status=summa.fit[,1],\n",
    "         anno=fit.cont$genes, side.main=\"ENTREZID\", folder=\"md\")\n",
    "\n",
    "# Test for DE relative to FC cut-off\n",
    "fit.treat <- treat(fit.cont, lfc=1)\n",
    "res.treat <- decideTests(fit.treat)\n",
    "summary(res.treat)\n",
    "\n",
    "topTable(fit.treat, coef=1, sort.by='p')\n",
    "\n",
    "par(mfrow=c(1,2))\n",
    "plotMD(fit.treat, coef=1, status=res.treat[,1], values=c(-1,1))\n",
    "plotMD(fit.treat, coef=2, status=res.treat[,2], values=c(-1,1))\n",
    "\n",
    "# Challenge. Relax the fc cut-off in a treat analysis to a \n",
    "# 50% change in fold-change and repeat analysis and plots\n",
    "log2(1.5)\n",
    "\n",
    "fit.treat <- treat(fit.cont, lfc=log2(1.5))\n",
    "res.treat <- decideTests(fit.treat)\n",
    "summary(res.treat)\n",
    "\n",
    "topTable(fit.treat, coef=1, sort.by='p')\n",
    "\n",
    "par(mfrow=c(1,2))\n",
    "plotMD(fit.treat, coef=1, status=res.treat[,1], values=c(-1,1))\n",
    "plotMD(fit.treat, coef=2, status=res.treat[,2], values=c(-1,1))\n",
    "\n",
    "# Let's make a final interactive MD plot of these results\n",
    "glMDPlot(fit.treat, coef=1, main=\"B.PregVsLac\",\n",
    "         counts=v$E, groups=group2, status=res.treat[,1],\n",
    "         anno=fit.cont$genes, side.main=\"ENTREZID\", folder=\"mdtreat\")\n",
    "\n",
    "# Gene set testing analysis\n",
    "# Let's start with an overlap test - goana\n",
    "go <- goana(fit.cont, coef=\"B.PregVsLac\", species=\"Mm\")\n",
    "topGO(go, n=10)\n",
    "\n",
    "colnames(seqdata)\n",
    "dim(seqdata)\n",
    "dim(fit.cont)\n",
    "\n",
    "# Let's take gene length into account\n",
    "# Need to match up the lengths from the original counts file\n",
    "\n",
    "m <- match(rownames(fit.cont), seqdata$EntrezGeneID)\n",
    "gene_length <- seqdata$Length[m]\n",
    "head(gene_length)\n",
    "\n",
    "go_length <- goana(fit.cont, coef=\"B.PregVsLac\", species=\"Mm\",\n",
    "                   covariate = gene_length)\n",
    "topGO(go_length, n=10)\n",
    "\n",
    "# Challenge\n",
    "# Repeat our goana ovelap analysis with gene length correction\n",
    "# For L.PregVsLac\n",
    "go_length_lum <- goana(fit.cont, coef=\"L.PregVsLac\", species=\"Mm\",\n",
    "                   covariate = gene_length)\n",
    "topGO(go_length_lum, n=10)\n",
    "\n",
    "# Loaded out MSigDB mouse sigantures into R \n",
    "# (these were downloaded with the data)\n",
    "load(\"data/mouse_c2_v5.rdata\")\n",
    "length(Mm.c2)\n",
    "names(Mm.c2)[1:5]\n",
    "Mm.c2$KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
    "\n",
    "# Matched these with our RNA-seq data\n",
    "c2.ind <- ids2indices(Mm.c2, rownames(v))\n",
    "length(c2.ind)\n",
    "c2.ind$KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
    "\n",
    "# Camera test in B.PregVsLac\n",
    "gst.camera <- camera(v, index=c2.ind, contrast=cont.matrix[,1])\n",
    "gst.camera[1:5,]\n",
    "\n",
    "table(gst.camera$FDR < 0.05)\n",
    "\n",
    "# Challenge\n",
    "# Run camera on c2 gene signatures for L.PregVsLac\n",
    "gst.camera_lum <- camera(v, index=c2.ind, contrast=cont.matrix2[,2])\n",
    "gst.camera_lum[1:5,]\n",
    "\n",
    "# Run camera on Hallmark MSigDB signature sets available \n",
    "# in mouse_H_v5.rdata\n",
    "load(\"data/mouse_H_v5.rdata\")\n",
    "length(Mm.H)\n",
    "names(Mm.H)[1:5]\n",
    "\n",
    "# Matched these with our RNA-seq data\n",
    "h.ind <- ids2indices(Mm.H, rownames(v))\n",
    "length(h.ind)\n",
    "\n",
    "# Camera test in B.PregVsLac\n",
    "gst.camera_hallmark <- camera(v, index=h.ind, contrast=cont.matrix[,1])\n",
    "gst.camera_hallmark[1:5,]\n",
    "\n",
    "table(gst.camera_hallmark$FDR < 0.05)\n",
    "\n",
    "# MYC is really interesting, so let's find if there are any other MYC\n",
    "# signatures coming up in our c2 results\n",
    "myc <- grep(\"MYC_\", names(c2.ind))\n",
    "names(c2.ind)[myc]\n",
    "\n",
    "myc.roast <- roast(v, index=c2.ind[myc], \n",
    "                   contrast = cont.matrix[,1], nrot=999)\n",
    "myc.roast[1:15,]\n",
    "\n",
    "grep(\"MYC_\", rownames(gst.camera))\n",
    "\n",
    "# Challenge: Can you test whether the MYC sigantures\n",
    "# are also DE in the L.PregvsLac?\n",
    "myc.roast_lum <- roast(v, index=c2.ind[myc], \n",
    "                   contrast = cont.matrix2[,2], nrot=999)\n",
    "myc.roast_lum[1:15,]\n",
    "\n",
    "# Can you test whether the WNT sigantures\n",
    "# are also DE in the B.PregvsLac?\n",
    "wnt <- grep(\"WNT_\", names(c2.ind))\n",
    "names(c2.ind)[wnt]\n",
    "\n",
    "wnt.roast <- roast(v, index=c2.ind[wnt], \n",
    "                   contrast = cont.matrix[,1], nrot=999)\n",
    "wnt.roast[1:5,]\n",
    "\n",
    "# Let's finish with a barcodeplot\n",
    "names(fit.cont)\n",
    "head(fit.cont$t)\n",
    "\n",
    "par(mfrow=c(1,1))\n",
    "barcodeplot(fit.cont$t[,1], \n",
    "            index=c2.ind[[\"MENSSEN_MYC_TARGETS\"]],\n",
    "            main=\"B.PregVsLac: MENSSEN_MYC_TARGETS (t-stats)\")\n",
    "\n",
    "# Final challenge:\n",
    "# Create a barcodeplot of MENSSEN_MYC_TARGETS for L.PregVsLac\n",
    "barcodeplot(fit.cont$t[,2], \n",
    "            index=c2.ind[[\"MENSSEN_MYC_TARGETS\"]],\n",
    "            main=\"L.PregVsLac: MENSSEN_MYC_TARGETS (t-stats)\")\n",
    "\n",
    "# Choose one of the WNT signalling pathway to barcodeplot\n",
    "par(mfrow=c(1,2))\n",
    "barcodeplot(fit.cont$t[,1], \n",
    "            index=c2.ind[[\"WILLERT_WNT_SIGNALING\"]],\n",
    "            main=\"B.PregVsLac: WILLERT_WNT_SIGNALING (t-stats)\")\n",
    "barcodeplot(fit.cont$t[,2], \n",
    "            index=c2.ind[[\"WILLERT_WNT_SIGNALING\"]],\n",
    "            main=\"L.PregVsLac: WILLERT_WNT_SIGNALING (t-stats)\")\n",
    "\n",
    "# Ugly plot of MYC and WNT on one figure\n",
    "# Hint: use index2\n",
    "par(mfrow=c(1,2))\n",
    "barcodeplot(fit.cont$t[,1], \n",
    "            index=c2.ind[[\"MENSSEN_MYC_TARGETS\"]],\n",
    "            index2=c2.ind[[\"WILLERT_WNT_SIGNALING\"]],\n",
    "            main=\"B.PregVsLac: MENSSEN_MYC_TARGETS & WILLERT_WNT_SIGNALING (t-stats)\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
